### PR TITLE
BCFR-967 - EVM Chain bindings for CR/CW - Basic support for method

### PR DESCRIPTION
**Pull request by pablolagreca**
Adding test cases for chain components using EVM generated bindings

[BCFR-967](https://smartcontract-it.atlassian.net/browse/BCFR-967)

### Requires
- https://github.com/smartcontractkit/chainlink-common/pull/829

[BCFR-967]: https://smartcontract-it.atlassian.net/browse/BCFR-967?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/services/relay/evm/evmtesting/run_tests.go [hunk 1]
```diff
@@ -33,194 +43,242 @@ func RunChainComponentsInLoopEvmTests[T TestingT[T]](t T, it ChainComponentsInte
...
-		reader := it.GetContractReader(t)
-
-		ctx := it.Helper.Context(t)
-		err := reader.Bind(ctx, []clcommontypes.BoundContract{{Name: AnyContractName, Address: addr.Hex()}})
-
-		require.ErrorIs(t, err, read.NoContractExistsError{Address: addr})
-	})
+	testCases := []Testcase[T]{
+		{
+			Name: DynamicTypedTopicsFilterAndCorrectReturn,
```
Comment (ilija42) [inline core/services/relay/evm/evmtesting/run_tests.go]: Is this leakage from [BCFR-967?](https://github.com/smartcontractkit/chainlink-common/pull/829) ?

Comment (ilija42) [reply core/services/relay/evm/evmtesting/run_tests.go]: Makes it hard to review

File: core/services/relay/evm/evmtesting/chain_components_interface_tester.go [hunk 2]
```diff
@@ -105,17 +109,99 @@ func (it *EVMChainComponentsInterfaceTester[T]) Setup(t T) {
...
+	it.GetContractReader(t)
+	it.txm = it.Helper.TXM(t, it.client)
+
+	it.chainWriterConfig = it.chainWriterConfigSupplier(t)
 
+	it.deployNewContracts(t)
+	it.dirtyConfig = false
+}
+
+func (it *EVMChainComponentsInterfaceTester[T]) getChainWriterConfig(t T) types.ChainWriterConfig {
```
Comment (ilija42) [inline core/services/relay/evm/evmtesting/chain_components_interface_tester.go:123]: why extract chainWriterConfig, but not the CR one

File: core/services/relay/evm/bindings/chain_reader_tester.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+	ContractReader types.ContractReader
+	ChainWriter    types.ChainWriter
+}
+
+type AddTestStructInput struct {
+	Field               int32
+	DifferentField      string
+	OracleId            uint8
+	OracleIds           [32]uint8
+	Account             common.Address
```
Comment (ilija42) [inline core/services/relay/evm/bindings/chain_reader_tester.go]: the tool should generate string types for address types

**Review by ilija42 — DISMISSED** [review]

File: core/services/relay/evm/evmtesting/run_tests.go [hunk 4]
```diff
@@ -21,6 +21,17 @@ import (
...
 
+const (
+	QueryKeyFilterOnDataWordsWithValueComparator                             = "Filtering can be done on data words using value comparator"
+	QueryKeyOnDataWordsWithValueComparatorOnNestedField                      = "Filtering can be done on data words using value comparator on a nested field"
+	QueryKeyFilterOnDataWordsWithValueComparatorOnDynamicField               = "Filtering can be done on data words using value comparator on field that follows a dynamic field"
+	QueryKeyFilteringOnDataWordsUsingValueComparatorsOnFieldsWithManualIndex = "Filtering can be done on data words using value comparators on fields that require manual index input"
+	DynamicTypedTopicsFilterAndCorrectReturn                                 = "Dynamically typed topics can be used to filter and have type correct in return"
+	MultipleTopicCanFilterTogether                                           = "Multiple topics can filter together"
+	FilteringCanBeDoneOnHashedIndexedTopics                                  = "Filtering can be done on indexed topics that get hashed"
+	BindReturnsErrorOnMissingContractAtAddress                               = "Bind returns error on missing contract at address"
```
Comment (jmank88) [inline core/services/relay/evm/evmtesting/run_tests.go]: This names are pretty hard to read. What is the advantage of using constants over inline literals?

File: core/services/relay/evm/evmtesting/run_tests.go [hunk 5]
```diff
@@ -34,237 +45,291 @@ func RunChainComponentsInLoopEvmTests[T TestingT[T]](t T, it ChainComponentsInte
...
+					sequences, err := cr.QueryKey(ctx, boundContract, query.KeyFilter{Key: staticBytesEventName, Expressions: []query.Expression{
+						query.Comparator("msgTransmitterEvent",
+							primitives.ValueComparator{
+								Value:    dataWordOnChainValueToQuery,
+								Operator: primitives.Eq,
+							}),
+					},
+					}, query.LimitAndSort{}, eventResAsStruct{})
+					return err == nil && len(sequences) == 1 && reflect.DeepEqual(wrapExpectedRes, sequences[0].Data)
+				}, it.MaxWaitTimeForEvents(), time.Millisecond*10)
```
Comment (jmank88) [inline core/services/relay/evm/evmtesting/run_tests.go:332]: Pre-existing, but 10ms seems unnecessarily fast:
```suggestion
				}, it.MaxWaitTimeForEvents(), 100*time.Millisecond)
```

**Review by EasterTheBunny — DISMISSED** [review]

**Review by ilija42 — DISMISSED** [review]

**Review by ilija42 — APPROVED** [review]

**Review by jmank88 — APPROVED** [review]

**Pull request merged**